docs/dev/lwt-implementation.md

```diff
@@ -0,0 +1,166 @@
+# Lightweight Transactions in Scylla
+
+The Lightweight Transactions (LWTs) are used in cases where 'linearizable' consistency model is required. The LWT implementation in Scylla relies on Paxos algorithm to achieve it.
+
+The scope of Paxos is partition key. Each LWT request works with only one partition key, accessing many partitions in a single LWT request is not supported. LWT requests to different partitions are completely independent of each other and are executed in parallel. LWT requests to the same partition wait for each other if the same node handles them. Concurrent LWTs to the same partition from different nodes run in parallel. The semantics of such execution is specified by Paxos. Note that the same rules apply for requests to different clustering keys in the same partition.
+
+A Scylla node maintains a local paxos state for each partition key in system.paxos table. This table uses the partition key from the base table as its own partition key. This ensures that all reads and writes to this table are local to the shard, handling LWT request for that partition key. This is important to avoid cross-shard accesses in cases when both system.paxos and the base table need to be accessed in the same request.
+
+The goal of Paxos algorithm is to ensure consistency. If two different users observe the same state at the same time and want to make their own modifications, it must be guaranteed that either the first or the second modification goes first, and the other modification sees the effects of the first one. This requirement is known as linearizable consistency model. Scylla provides this consistency model only for LWT requests. If one user runs LWT requests and another accesses the same data through regular quorum reads and writes, linearizability is not guaranteed.
+
+Modifications are called proposals in Paxos. Each proposal contains a mutation and a timestamp, which is called a ballot. Any node (LWT coordinator) can act as a 'proposer' -- by sending a new proposal to the replica set it _proposes_ its own new state (represented by the mutation inside the proposal) for the replica set to commit as its new acknowledged state.
+
+The LWT entry point is `storage_proxy::cas` method. It takes the following parameters:
```

Comment:
I think you forgot the most important parameter - the "cas_request" object which should be subclassed to implement the actual logic of the what the LWT request writes based on what read_command does. How to subclass cas_request, how its apply() works (it can be run multiple times, it gets a timestamp, etc.) should be explained somewhere.

Another thing you didn't mention is the fact that you **must** run storage_proxy::cas() on a specific shard, `storage_proxy::cas_shard(*_schema, token)`. If you try to do it from any other shard, the request is rejected.